File: app/src/main/java/com/codemybrainsout/imageviewer/view/detail/DetailActivity.java
package com.codemybrainsout.imageviewer.view.detail;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.codemybrainsout.imageviewer.R;
import com.codemybrainsout.imageviewer.api.PhotoService;
import com.codemybrainsout.imageviewer.custom.CircleImageView;
import com.codemybrainsout.imageviewer.custom.ExifDialog;
import com.codemybrainsout.imageviewer.model.Exif;
import com.codemybrainsout.imageviewer.model.Photo;
import com.codemybrainsout.imageviewer.model.User;
import com.codemybrainsout.imageviewer.service.FileDownloadService;
import com.codemybrainsout.imageviewer.utility.Extra;
import com.codemybrainsout.imageviewer.view.base.BaseActivity;
import com.codemybrainsout.imageviewer.view.user.UserActivity;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import it.sephiroth.android.library.imagezoom.ImageViewTouch;
import pub.devrel.easypermissions.AppSettingsDialog;
import pub.devrel.easypermissions.EasyPermissions;

public class DetailActivity extends BaseActivity implements DetailContract.View, EasyPermissions.PermissionCallbacks {

    private static final String EXTRA_PHOTO = "photo_extra";
    private static final String TAG = DetailActivity.class.getSimpleName();

    @BindView(R.id.activity_detail_user_IV)
    CircleImageView activityDetailUserIV;
    @BindView(R.id.activity_detail_user_TV)
    TextView activityDetailUserTV;
    @BindView(R.id.activity_detail_toolbar)
    Toolbar activityDetailToolbar;
    @BindView(R.id.activity_detail_IVT)
    ImageViewTouch activityDetailIVT;
    @BindView(R.id.activity_detail_user_LL)
    LinearLayout activityDetailUserLL;
    @BindView(R.id.parent)
    RelativeLayout parent;
    @BindView(R.id.activity_detail_set_wallpaper_IV)
    ImageView activityDetailSetWallpaperIV;
    @BindView(R.id.activity_detail_set_wallpaper)
    LinearLayout activityDetailSetWallpaper;
    @BindView(R.id.activity_detail_download)
    LinearLayout activityDetailDownload;
    @BindView(R.id.activity_detail_info_LL)
    LinearLayout activityDetailInfoLL;

    private Photo photo;

    @Inject
    public PhotoService photoService;

    public DetailPresenter mPresenter;

    private int PERMISSIONS = 100;
    private String[] per = {android.Manifest.permission.READ_EXTERNAL_STORAGE, android.Manifest.permission.WRITE_EXTERNAL_STORAGE};

    public static Intent getDetailActivityIntent(Context context, Photo photo) {
        Intent intent = new Intent(context, DetailActivity.class);
        intent.putExtra(DetailActivity.EXTRA_PHOTO, photo);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail);
        ButterKnife.bind(this);
        setToolbar();

        DetailComponent detailComponent = DaggerDetailComponent.builder()
                .appComponent(getAppComponent())
                .detailModule(new DetailModule(this))
                .build();
        detailComponent.inject(this);

        if (getIntent().hasExtra(EXTRA_PHOTO)) {
            photo = (Photo) getIntent().getSerializableExtra(EXTRA_PHOTO);
            mPresenter = new DetailPresenter(photoService, this, photo);
            mPresenter.loadPhoto();
        }
    }

    @Override
    public void showLoading() {
        addProgressLayout(this, parent);
    }

    @Override
    public void hideLoading() {
        removeProgressLayout(parent);
    }

    @Override
    public void showPhoto(final Photo photo) {

        this.photo = photo;
        final User user = photo.getUser();

        if (!TextUtils.isEmpty(user.getProfileImage().getMedium())) {
            Glide.with(this)
                    .load(user.getProfileImage().getMedium())
                    .into(activityDetailUserIV);
        }

        activityDetailUserTV.setText(user.getName());
        activityDetailUserLL.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mPresenter.openUserDetails(user);
            }
        });

        if (!TextUtils.isEmpty(photo.getUrls().getRegular())) {
            Glide.with(this)
                    .load(photo.getUrls().getRegular())
                    .placeholder(Color.parseColor(photo.getColor()))
                    .into(activityDetailIVT);
        }

        //photo.getLocation();
        //photo.getLikes();
        //photo.getDownloads();
        activityDetailInfoLL.setVisibility(View.VISIBLE);

    }

    @OnClick(R.id.activity_detail_set_wallpaper)
    public void setAsWallpaper() {
        if (EasyPermissions.hasPermissions(this, per)) {
            if (!TextUtils.isEmpty(photo.getUrls().getFull())) {
                activityDetailSetWallpaperIV.performClick();
                mPresenter.setAsWallpaper(this, photo.getUrls().getFull());
            }
        } else {
            request();
        }
    }

    @OnClick(R.id.activity_detail_download)
    public void downloadImage() {
        if (EasyPermissions.hasPermissions(this, per)) {
            if (!TextUtils.isEmpty(photo.getUrls().getFull())) {
                mPresenter.downloadPhoto(this, photo.getUrls().getFull());
            }
        } else {
            request();
        }
    }

    @OnClick(R.id.activity_detail_info_LL)
    public void openInfo() {
        Exif exif = photo.getExif();
        if (exif != null) {
            if (photo.getDownloads() != null) {
                exif.setDownloads(photo.getDownloads());
            }
            if (photo.getLikes() != null) {
                exif.setLikes(photo.getLikes());
            }
            mPresenter.showExif(exif);
        }
    }

    @Override
    public void showExif(Exif exif) {
        //show exif view
        new ExifDialog(this, exif).show();
    }

    @Override
    public void showUserDetails(User user) {
        Intent intent = UserActivity.getUserActivityIntent(this, user);
        startActivity(intent);
    }

    private BroadcastReceiver downloadReceiver = new BroadcastReceiver() {

        // Called when the broadcast is received
        @Override
        public void onReceive(Context context, Intent intent) {

            boolean success = intent.getBooleanExtra(Extra.SUCCESS, false);
            boolean wallpaper = intent.getBooleanExtra(Extra.WALLPAPER, false);

            if (success) {

                if (wallpaper) {
                    showErrorSnackbar(parent, R.string.wallpaper_success);
                } else {
                    showErrorSnackbar(parent, R.string.download_success);
                }

            } else {

                showErrorSnackbar(parent, R.string.error);

            }

        }
    };

    @Override
    public void onResume() {
        super.onResume();
        //set filter to only when download is complete and register broadcast receiver
        IntentFilter filter = new IntentFilter(FileDownloadService.ACTION_DOWNLOAD_COMPLETE);
        LocalBroadcastManager.getInstance(this).registerReceiver(downloadReceiver, filter);
    }

    @Override
    protected void onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(downloadReceiver);
        super.onPause();
    }

    @Override
    public void showError(String s) {
        showErrorSnackbar(parent, s);
    }

    @Override
    public void setToolbar() {
        setSupportActionBar(activityDetailToolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowTitleEnabled(false);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void request() {
        EasyPermissions.requestPermissions(this, getString(R.string.permission_rationale), PERMISSIONS, per);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        // Forward results to EasyPermissions
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
    }

    @Override
    public void onPermissionsGranted(int requestCode, List<String> perms) {
        if (!EasyPermissions.hasPermissions(this, per)) {
            request();
        }
    }

    @Override
    public void onPermissionsDenied(int requestCode, List<String> perms) {
        if (EasyPermissions.somePermissionPermanentlyDenied(this, perms)) {
            new AppSettingsDialog.Builder(this).build().show();
        }
    }
}

File: app/src/main/java/com/codemybrainsout/imageviewer/view/detail/DetailActivity.kt
package com.codemybrainsout.imageviewer.view.detail

import android.Manifest
import android.Manifest.permission.READ_EXTERNAL_STORAGE
import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.graphics.Color.parseColor
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.View
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.bumptech.glide.Glide
import com.codemybrainsout.imageviewer.R
import com.codemybrainsout.imageviewer.custom.ExifDialog
import com.codemybrainsout.imageviewer.databinding.ActivityDetailBinding
import com.codemybrainsout.imageviewer.extensions.addProgressLayout
import com.codemybrainsout.imageviewer.extensions.removeProgressLayout
import com.codemybrainsout.imageviewer.extensions.showErrorSnackbar
import com.codemybrainsout.imageviewer.extensions.viewBinding
import com.codemybrainsout.imageviewer.model.Exif
import com.codemybrainsout.imageviewer.model.Photo
import com.codemybrainsout.imageviewer.model.User
import com.codemybrainsout.imageviewer.service.FileDownloadService
import com.codemybrainsout.imageviewer.utility.Extra
import com.codemybrainsout.imageviewer.view.base.BaseActivity
import com.codemybrainsout.imageviewer.view.user.UserActivity
import com.vmadalin.easypermissions.EasyPermissions
import com.vmadalin.easypermissions.annotations.AfterPermissionGranted
import com.vmadalin.easypermissions.dialogs.SettingsDialog
import com.vmadalin.easypermissions.models.PermissionRequest
import javax.inject.Inject

class DetailActivity : BaseActivity(), DetailContract.View, EasyPermissions.PermissionCallbacks {

    private val binding by viewBinding(ActivityDetailBinding::inflate)

    @Inject
    lateinit var mPresenter: DetailPresenter

    private var photo: Photo? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (intent.hasExtra(EXTRA_PHOTO)) {
            photo = intent.getParcelableExtra(EXTRA_PHOTO)
            loadDetails()
        }

        binding.apply {
            layoutSetWallpaper.setOnClickListener { setAsWallpaper() }
            layoutDownload.setOnClickListener { downloadImage() }
            layoutInfo.setOnClickListener { openInfo() }
            toolbar.setNavigationOnClickListener { onBackPressed() }
        }
    }

    private fun loadDetails() {
        photo?.id?.let { mPresenter.loadPhoto(it) }
    }

    override fun showLoading() {
        binding.root.addProgressLayout(this)
    }

    override fun hideLoading() {
        binding.root.removeProgressLayout()
    }

    override fun showPhoto(photo: Photo) {
        this.photo = photo
        photo.user?.let { user ->
            if (user.profileImage?.medium.isNullOrEmpty().not()) {
                Glide.with(this)
                    .load(user.profileImage?.medium)
                    .into(binding.imageViewUser)
            }
            binding.textViewUser.text = user.name
            binding.layoutUser.setOnClickListener {
                mPresenter.openUserDetails(user)
            }
        }

        val color = parseColor(photo.color)
        if (photo.urls?.regular.isNullOrEmpty().not()) {
            Glide.with(this)
                .load(photo.urls?.regular)
                .placeholder(ColorDrawable(color))
                .into(binding.imageViewTouch)
        }

        binding.layoutInfo.visibility = View.VISIBLE
    }

    @AfterPermissionGranted(PERMISSIONS_WALLPAPER)
    private fun setAsWallpaper() {
        if (EasyPermissions.hasPermissions(this, READ_EXTERNAL_STORAGE, WRITE_EXTERNAL_STORAGE)) {
            photo?.urls?.full?.let {
                if (it.isNotEmpty()) {
                    binding.imageViewSetWallpaper.performClick()
                    mPresenter.setAsWallpaper(this, it)
                }
            }
        } else {
            request(PERMISSIONS_WALLPAPER)
        }
    }

    @AfterPermissionGranted(PERMISSIONS_DOWNLOAD)
    private fun downloadImage() {
        if (EasyPermissions.hasPermissions(this, READ_EXTERNAL_STORAGE, WRITE_EXTERNAL_STORAGE)) {
            photo?.urls?.full?.let {
                if (it.isNotEmpty()) {
                    mPresenter.downloadPhoto(this, it)
                }
            }
        } else {
            request(PERMISSIONS_DOWNLOAD)
        }
    }

    private fun openInfo() {
        val exif: Exif? = photo?.exif
        if (exif != null) {
            if (photo?.downloads != null) {
                exif.downloads = photo?.downloads
            }
            if (photo?.likes != null) {
                exif.likes = photo?.likes
            }
            mPresenter.showExif(exif)
        }
    }

    override fun showExif(exif: Exif?) {
        //show exif view
        exif?.let { ExifDialog(this, it).show() }
    }

    override fun showUserDetails(user: User?) {
        val intent: Intent = UserActivity.getUserActivityIntent(this, user)
        startActivity(intent)
    }

    private val downloadReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        // Called when the broadcast is received
        override fun onReceive(context: Context, intent: Intent) {
            val success: Boolean = intent.getBooleanExtra(Extra.SUCCESS, false)
            val wallpaper: Boolean = intent.getBooleanExtra(Extra.WALLPAPER, false)
            if (success) {
                if (wallpaper) {
                    binding.root.showErrorSnackbar(R.string.wallpaper_success)
                } else {
                    binding.root.showErrorSnackbar(R.string.download_success)
                }
            } else {
                binding.root.showErrorSnackbar(R.string.error)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        //set filter to only when download is complete and register broadcast receiver
        val filter = IntentFilter(FileDownloadService.ACTION_DOWNLOAD_COMPLETE)
        LocalBroadcastManager.getInstance(this).registerReceiver(downloadReceiver, filter)
    }

    override fun onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(downloadReceiver)
        super.onPause()
    }

    override fun showError(s: String?) {
        binding.root.showErrorSnackbar(s ?: "")
    }

    private fun request(requestCode: Int) {
        if (!EasyPermissions.hasPermissions(this, READ_EXTERNAL_STORAGE, WRITE_EXTERNAL_STORAGE)) {
            val request = PermissionRequest.Builder(this)
                .code(requestCode)
                .perms(arrayOf(READ_EXTERNAL_STORAGE, WRITE_EXTERNAL_STORAGE))
                .rationale(R.string.permission_rationale)
                .build()
            EasyPermissions.requestPermissions(this, request)
        }
    }

    override fun onPermissionsDenied(requestCode: Int, perms: List<String>) {
        if (EasyPermissions.somePermissionPermanentlyDenied(this, perms)) {
            SettingsDialog.Builder(this).build().show()
        }
    }

    override fun onPermissionsGranted(requestCode: Int, perms: List<String>) {
        // do nothing
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        // EasyPermissions handles the request result.
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this)
    }

    companion object {
        private const val PERMISSIONS_WALLPAPER = 100
        private const val PERMISSIONS_DOWNLOAD = 100
        private const val EXTRA_PHOTO = "photo_extra"
        private val TAG = DetailActivity::class.java.simpleName
        fun getDetailActivityIntent(context: Context?, photo: Photo?): Intent {
            val intent = Intent(context, DetailActivity::class.java)
            intent.putExtra(EXTRA_PHOTO, photo)
            return intent
        }
    }
}